### ISSUE TITLE
Nightly regression test

### ISSUE THREAD

**Issue by pocke** [rubocop-hq/rubocop#6805]
I'd like to introduce a nightly regression test for RuboCop to reduce RuboCop's bugs.
RuboCop has two types of bug. First, it is called false positive/negative. Second, it is error, that is raised by Ruby process.
it cannot (or really hard) find the first type of bug automatically. Because false positive/negative are decided by human. But it can find the second type of bug automatically.
So the regression test focuses the second.

How to test
===

The regression test executes RuboCop on many repositories with many configurations.

I'd like to use OSS Ruby projects for the test. We can find many Ruby codes from GitHub. There are many coding styles, so they may have unexpected code for cops.

We already have RSpec test cases, but they cover only a few code styles. There are many styles in the real world. It is a weak point of the existing test cases.
I think the regression test is helpful to reduce the weak point. 

PoC
===

I have a proof of concept implementation. https://github.com/pocke/rubocop-regression-test
It generates many RuboCop's configurations with changes `EnforcedStyle`, and executes RuboCop with the configurations. It reports errors if they occur.

Usage is simple.

```bash
# RuboCop will be executed in https://github.com/ruby/ruby repository.
$ ruby main.rb rails/rails

# RuboCop will be executed in the specified repositories by https://github.com/pocke/rubocop-regression-test/blob/22b241d0da8081bff4791d5bdc3def7426745475/lib/cli.rb#L2-L16
$ ruby main.rb
```

I found bugs with the PoC, such as #6803 and #6802.

## Problems

I do not have any place to run the regression test. I execute it on my local computer, but it is not good for production.
I think a nightly test is best. Because it takes a lot of time. I guess it will take several hours. So it will block merging pull request if it is enabled for each pull request.
I'ld like to use a CI service for it, such as CircleCI. But I'm not sure we can use CircleCI, because I guess it will use build time for rubocop-hq/rubocop's test.
I'm not familar with it. Let me know if you know a good platform for the test.

**Comment by Drenmi** [comment]
Ping @bquorning.

**Comment by koic** [comment]
This looks like a good idea to me. Unfortunately I'm not familiar with it either.
@bquorning Are you interested in CircleCI setting for this nightly build?

**Comment by bquorning** [comment]
A nightly test is a good idea.

An example of how to configure CircleCI can is the [workflow:daily task in circleci-ruby-snapshot repository](https://github.com/rubocop-hq/circleci-ruby-snapshot-image/blob/4f867370259263fd75bd831a66bfec339b1fe1d1/.circleci/config.yml#L51-L73) which you can use as a starting point. Ping me if you run into problems – I’m happy to help.

**Comment by bbatsov** [comment]
Seems like a pretty cool idea! Looking forward to seeing how this will work out in practice!

**Comment by pocke** [comment]
Sorry for no response.
It already runs on CircleCI. https://circleci.com/gh/pocke/rubocop-regression-test
RuboCop's Circle CI configuration is really useful for me. Thank you @bquorning 




It works, but it has limitations.

* It tests only trick codes. 
  * https://github.com/pocke/rubocop-regression-test/blob/6730a980980f75dfa342cb1699345d25f678841f/lib/cli.rb#L2-L18
  * We need to fix existing bugs to use other repositories for the test. Because RuboCop fails on these repositories.
* Error report is not helpful
  * If the test fails, it notifies only to me.
* Difficult to reproduce
  * If a bug found by the regression test, we need to reproduce the bug in a local computer.
  * But the configuration file, which is used in the failed test, does not exist.
  * https://github.com/pocke/rubocop-regression-test/issues/3

**Comment by Earlopain** [comment]
Gonna revive this one. In addition to nightly tests, it would also make sense to run RuboCop against some repos and show how offenses change because of a PR. For example with https://github.com/rubocop/rubocop/pull/13729 this change would hopefully have stood out:

```rb
internal_execute("SELECT setval(#{quote(quoted_sequence)}, #{max_pk || minvalue}, #{max_pk ? true : false})", "SCHEMA")
# =>
internal_execute("SELECT setval(#{quote(quoted_sequence)}, #{max_pk || minvalue}, #{max_pk || false})", "SCHEMA")
```

Which quite drastically changes the executed sql

**Comment by viralpraxis** [comment]
I made this [tool](https://github.com/viralpraxis/rubocop-nightly) as an extended version of Pocke's work. I didn't spend much time with it but it exposed ~80 bugs (I fixed ~40 of them: https://github.com/rubocop/rubocop/pulls/viralpraxis). I believe its an extended version since it explores configuration state space more deeply and supports multiple sources (rubygems latest API, rubygems-mirror, git), and supports ~~extensions~~ plugins.

**Comment by Earlopain** [comment]
Ah, nice. Have have something [similar ](https://gist.github.com/Earlopain/b307c9d8b243023c90045adde75ef87a)and fixed a bunch of stuff with it too. Errors only as well, which wouldn't have done much in this specific case.

**Comment by bbatsov** [comment]
Yeah, I agree that something like this would be valuable to have.

**Comment by viralpraxis** [comment]
fyi @Earlopain 

I've improved [the tool](https://github.com/viralpraxis/rubocop-nightly) so it's now capable of finding the difference of offenses produced by two RuboCop revisions against given repository:

```console
$ bin/rubocop-nightly compare --from 53e5d198f --to master --source https://github.com/rails/rails.git:main

<snip>

Removed offenses: (10):
[Lint/LiteralAsCondition] activerecord/lib/active_record/connection_adapters/mysql/schema_creation.rb:43:62: Literal `"UNIQUE"` appeared as a condition.
[Style/RedundantCondition] activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb:359:104: Use double pipes `||` instead.
[Lint/LiteralAsCondition] activerecord/lib/active_record/relation/predicate_builder.rb:59:76: Literal `:eq` appeared as a condition.
[Style/RedundantCondition] activerecord/lib/active_record/test_fixtures.rb:310:9: Use double pipes `||` instead.
[Lint/ShadowingOuterLocalVariable] activerecord/test/cases/associations/eager_test.rb:1661:26: Shadowing outer local variable - `post`.
[Lint/ShadowingOuterLocalVariable] activerecord/test/cases/associations/eager_test.rb:1685:32: Shadowing outer local variable - `comment`.
[Lint/ShadowingOuterLocalVariable] activerecord/test/cases/migration/compatibility_test.rb:271:53: Shadowing outer local variable - `column`.
[Lint/ShadowingOuterLocalVariable] activerecord/test/cases/migration/exclusion_constraint_test.rb:57:56: Shadowing outer local variable - `constraint`.
[Lint/ShadowingOuterLocalVariable] activerecord/test/cases/migration/unique_constraint_test.rb:54:53: Shadowing outer local variable - `constraint`.
[Lint/ShadowingOuterLocalVariable] activerecord/test/cases/migration/unique_constraint_test.rb:62:53: Shadowing outer local variable - `constraint`.

New offenses: (18):
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/associations/collection_association.rb:70:32: Shadowing outer local variable - `primary_key`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/attribute_methods/composite_primary_key.rb:29:44: Shadowing outer local variable - `value`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/autosave_association.rb:521:49: Shadowing outer local variable - `key`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:530:45: Shadowing outer local variable - `row`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:531:50: Shadowing outer local variable - `row`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:1040:41: Shadowing outer local variable - `sql`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:475:45: Shadowing outer local variable - `row`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:476:50: Shadowing outer local variable - `row`.
[Style/RedundantParentheses] activerecord/lib/active_record/encryption/encrypted_attribute_type.rb:67:38: Don't use parentheses around a method argument.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/relation/query_methods.rb:1941:36: Shadowing outer local variable - `value`.
[Naming/PredicateName] activerecord/lib/active_record/relation/query_methods.rb:2051:11: Rename `does_not_support_reverse?` to `not_support_reverse?`.
[Lint/ShadowingOuterLocalVariable] activerecord/lib/active_record/tasks/database_tasks.rb:252:50: Shadowing outer local variable - `db_configs`.
[Naming/PredicateName] activerecord/lib/arel/predications.rb:147:9: Rename `does_not_match` to `not_match?`.
[Naming/PredicateName] activerecord/lib/arel/predications.rb:151:9: Rename `does_not_match_regexp` to `not_match_regexp?`.
[Naming/PredicateName] activerecord/lib/arel/predications.rb:155:9: Rename `does_not_match_any` to `not_match_any?`.
[Naming/PredicateName] activerecord/lib/arel/predications.rb:159:9: Rename `does_not_match_all` to `not_match_all?`.
[Style/RedundantParentheses] activerecord/test/cases/scoping/named_scoping_test.rb:92:18: Don't use parentheses around a method argument.
[Style/RedundantParentheses] activerecord/test/cases/scoping/named_scoping_test.rb:93:18: Don't use parentheses around a method argument.
```

**Comment by Starlexxx** [comment]
@bquorning The recent #15590 got me thinking about this again. That loop needed `EnforcedStyle: with_fixed_indentation` on one cop plus a non-default boolean on a second, disabled-by-default one. A scan with default configs can't hit that, and neither can sweeping one option at a time. The bugs that survive to release tend to live in config combinations.

So I built a tool that goes after exactly that: https://github.com/Starlexxx/rubocop-fuzz

It runs `rubocop -A` over a corpus of popular gems with generated configs: the defaults, every single-option deviation mined from `config/default.yml`, and option combinations for clusters of cops that read each other's config (found by grepping cop sources for `config.for_cop` and `cop_enabled?`, an idea borrowed from rubocop-nightly's DependenciesMiner). Each run checks for cop errors, infinite correction loops, output that no longer parses after correction, and corrections that show up again on a second pass. A delta-debugging step then shrinks each finding to a minimal source file and the smallest config that still reproduces it. Its own regression test is rediscovering #15590 on the revision before the fix.

What it has found on master so far:

* three infinite correction loops reachable with near-default config, fixed in #15597, #15598 and #15599;
* `Layout/IndentationWidth` raising `end_pos must not be less than begin_pos` whenever `Layout/IndentationStyle` is set to `tabs`;
* `Naming/BlockForwarding` and `Style/MethodDefParentheses` together rewriting `def each &block` into `def each(&))`, which does not parse;
* a tail of loop and broken-correction combinations I'm still working through.

This builds on rubocop-nightly by @viralpraxis and rubocop-regression-test by @pocke. I'll keep turning findings into issues and PRs; if there's interest in making something like this part of RuboCop's own process, glad to help.

**Comment by viralpraxis** [comment]
@Starlexxx two notes from my side:

1. `rubocop-nightly` is definitely good at finding bugs -- I'd say out of my [120 merged PRs](https://github.com/rubocop/rubocop/pulls?q=is%3Apr+author%3Aviralpraxis+is%3Aclosed) half of them were fixing bugs found by the tool. When I lost courage I still had ~40 findings left without patches.
2. I have a lot of `rubocop-nightly`'s enhancements/bugfixes unpublished (mostly related to performance, e.g. configuration permutation improvements) -- but given that I made this project in pre-claude era it might be that you already have them in place

**Comment by Starlexxx** [comment]
Thanks for the notes. I should run rubocop-nightly against current master myself and see what it still finds. To be fair, the only reason I started my own thing was that nightly looked abandoned when I was trying to reproduce bug, and I needed the cross-cop config part badly

**Comment by viralpraxis** [comment]
Let's see if maintainers are interested in this.

I believe it all makes sense since 1) practically all the work is already done 2) it's an easy way to improve DX. How many times have we released a new version with a bug that would have been discovered instantly, only to have to wait a week or smth for a new release?

I think we might start with a nightly CI job that runs all configuration permutations against e.g. the top-10 gems (including Rails subgems). `rubocop-nightly` (and perhaps @Starlexxx's `rubocop-fuzz` -- I haven't checked it yet) is already capable of doing this.

What do you think? @bbatsov @koic
